downstairs/src/extent_inner_raw.rs

```diff
@@ -0,0 +1,1757 @@
+                    );
+                    bail!(
+                        "Open of {path:?} for extent#{extent_number} \
+                         returned: {e}",
+                    );
+                }
+                Ok(f) => {
+                    let cur_size = f.metadata().unwrap().len();
+                    if size != cur_size {
+                        bail!(
+                            "File size {size:?} does not match \
+                             expected {cur_size:?}",
+                        );
+                    }
+                    f
+                }
+            };
+
+        // Just in case, let's be very sure that the file on disk is what it
+        // should be
+        if !read_only {
+            if let Err(e) = file.sync_all() {
+                return Err(CrucibleError::IoError(format!(
+                    "extent {extent_number}:
+                 fsync 1 failure during initial rehash: {e:?}",
+                ))
+                .into());
+            }
+        }
+
+        // Read the active context bitpacked array and metadata.  The former is
+        // only valid if `dirty` is false in the metadata, but that's the most
+        // common case, so we'll optimize for it.
+        let mut active_context_buf = vec![0u8; (bcount as usize + 7) / 8];
+        let mut meta_buf = [0u8; BLOCK_META_SIZE_BYTES as usize];
+        let mut iovecs = [
+            IoSliceMut::new(&mut active_context_buf),
+            IoSliceMut::new(&mut meta_buf),
+        ];
+        nix::sys::uio::preadv(
+            file.as_raw_fd(),
+            &mut iovecs,
+            (bcount * (def.block_size() + BLOCK_CONTEXT_SLOT_SIZE_BYTES * 2))
+                as i64,
+        )
+        .map_err(|e| CrucibleError::IoError(e.to_string()))?;
+
+        let dirty = match meta_buf[0] {
+            0 => false,
+            1 => true,
+            i => bail!("invalid dirty value: {i}"),
+        };
+
+        // If the file is dirty, then we have to recompute which context slot is
+        // active for every block.  This is slow, but can't be avoided; we
+        // closed the file without a flush so we can't be confident about the
+        // data that was on disk.
+        let active_context = if !dirty {
+            // Easy case first: if it's **not** dirty, then just assign active
+            // slots based on the bitpacked active context buffer from the file.
+            let mut active_context = vec![];
+            for b in active_context_buf {
+                // Unpack bits from each byte
+                for i in 0..8 {
+                    active_context.push(if b & (1 << i) == 0 {
+                        ContextSlot::A
+                    } else {
+                        ContextSlot::B
+                    });
+                }
+            }
+            // It's possible that block count isn't a multiple of 8; in that
+            // case, shrink down the active context array.
```

Comment:
We could assert this, what do you think?

Reply:
No, having a block count that isn't a multiple of 8 is totally fine – for example, all of our tests call `new_region_options`, which builds an extent with 10 blocks.